{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDA v2 - Example for subject 09CO022\n",
    "\n",
    "This notebook revisits a case looked at in version 1. Subject 09CO022 was a good example for CDA because they were known to have genomic, imaging and proteomic data. Several aspects of this case were looked at in v1 see [v1tests/09CO022 Example]((../v1tests/09CO022%20Example.ipynb)). This new notebook using CDA c2 revisits only those issues related to specimens and files. The issues in the original notebook related to Diagnosis and other phenotypic issues will be covered elsewhere.\n",
    "\n",
    "A diagram in the [v1 notebook](../v1tests/09CO022%20Example.ipynb) shows, independent of CDA, the hierarchy of specimens and sub-specimens and known images that were derived from that 09CO022 (a person). \n",
    "\n",
    "In the Genomic Data Commons portal we can see the three specimens that were used for genomic analysis. The ids for the specimens match those in blue in the diagram. Note that the specimen ids shown are those used by the TCGA/CPTAC2 projects, as opposed to the UUIDs within the GDC.\n",
    "\n",
    "A number of issues were identified in version 1. This notebook explores what has happened with those issues in version 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cdapython import Q, columns, unique_terms\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following query is what one would expect to return the subject shown above. We'll save the json result to a file. Browsing [that file](./query_results/09CO022.json) may be useful to following along with this notebook. It is too large to include in full here, but the key parts are shown."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting results from database\n",
      "\n",
      "Total execution time: 1362 ms\n",
      "\n",
      "QueryID: e85f9635-6494-49d0-b589-66c2ea1426bb\n",
      "Query: SELECT all_v2.* FROM gdc-bq-sample.integration.all_v2 AS all_v2, UNNEST(ResearchSubject) AS _ResearchSubject WHERE (_ResearchSubject.id = 'c5421e34-e5c7-4ba5-aed9-146a5575fd8d')\n",
      "Offset: 0\n",
      "Count: 1\n",
      "Total Row Count: 1\n",
      "More pages: False\n",
      "\n"
     ]
    }
   ],
   "source": [
    "q1 = Q('ResearchSubject.id = \"c5421e34-e5c7-4ba5-aed9-146a5575fd8d\"')\n",
    "r = q1.run(limit=2) \n",
    "r.sql\n",
    "print(r)\n",
    "j = json.dumps(r[0], indent=3)\n",
    "with open('query_results/09CO022.json', 'w') as f:\n",
    "    f.write(j)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Worth noting that this gave one result, as would be expected for a query specifying a single subject id.\n",
    "\n",
    "Does this return the expected specimens from the diagram? We need to look within each ResearchSubject?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specimens within ResearchSubject\n",
    "Looking at the specimens within Research Subject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id : c5421e34-e5c7-4ba5-aed9-146a5575fd8d\n",
      "identifier: GDC:c5421e34-e5c7-4ba5-aed9-146a5575fd8d\n",
      "associated_project : CPTAC-2\n",
      "primary_disease_type : Adenomas and Adenocarcinomas\n",
      "primary_disease_site : Colon\n",
      "Diagnosis has 1 items\n",
      "File has 30 items\n",
      "Specimen has 11 items\n",
      "________________________________________________________________________________\n",
      "id : 459e3b69-63d6-11e8-bcf1-0a2705229b82\n",
      "identifier: PDC:459e3b69-63d6-11e8-bcf1-0a2705229b82\n",
      "associated_project : CPTAC-2\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "primary_disease_site : Colon\n",
      "Diagnosis has 1 items\n",
      "File has 163 items\n",
      "Specimen has 5 items\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "for subj in r[0]['ResearchSubject']:\n",
    "    for k, v in subj.items():\n",
    "        if k == 'identifier':\n",
    "            print ('identifier: {}:{}'.format(v[0]['system'], v[0]['value']))\n",
    "        elif v.__class__.__name__ == 'list':\n",
    "            print ('{} has {} items'.format(k, len(v)))\n",
    "\n",
    "        else:\n",
    "            print ('{} : {}'.format(k, v))\n",
    "\n",
    "    print('_'*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's more that what we might expect. However, the following exploration of the specimens shows that the count of 11 is exactly the same hierarchy as in the GDC screenshot in the original notebook. Namely, two parent samples, sample->portion->analyte->aliquot hierarchy leading to three aliquots on which genomic sequencing was performed.\n",
    "\n",
    "#### First for the GDC \"Research Subject\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/forei/imfenvs/cda2/lib/python3.9/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>derived_from_specimen</th>\n",
       "      <th>specimen_type</th>\n",
       "      <th>source_material</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4591a53d-5668-4a70-b44b-e08a3d59267e</td>\n",
       "      <td>initial specimen</td>\n",
       "      <td>sample</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c53c4d60-2ddb-5da8-932e-00a86fa2347f</td>\n",
       "      <td>4591a53d-5668-4a70-b44b-e08a3d59267e</td>\n",
       "      <td>portion</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31075cfa-7aef-59f1-bf54-a1cddb5ee0fd</td>\n",
       "      <td>c53c4d60-2ddb-5da8-932e-00a86fa2347f</td>\n",
       "      <td>analyte</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0d8adcbf-13f0-48c3-83df-3fa205b79ae8</td>\n",
       "      <td>31075cfa-7aef-59f1-bf54-a1cddb5ee0fd</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d085ebd9-7605-54a0-abb9-10867f5fa1b1</td>\n",
       "      <td>4591a53d-5668-4a70-b44b-e08a3d59267e</td>\n",
       "      <td>portion</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>a31724b6-e550-552b-bd61-41341c534e28</td>\n",
       "      <td>d085ebd9-7605-54a0-abb9-10867f5fa1b1</td>\n",
       "      <td>analyte</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9250d96e-1cdc-4d68-8a56-f7b186a6fab5</td>\n",
       "      <td>a31724b6-e550-552b-bd61-41341c534e28</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>b12c257d-7409-4858-9384-c430929a075a</td>\n",
       "      <td>initial specimen</td>\n",
       "      <td>sample</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>702d7ba0-9558-5b2d-af4d-cd797485b8c1</td>\n",
       "      <td>b12c257d-7409-4858-9384-c430929a075a</td>\n",
       "      <td>portion</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>f0003f0a-07ea-548e-b1f7-7e6d1b27d47a</td>\n",
       "      <td>702d7ba0-9558-5b2d-af4d-cd797485b8c1</td>\n",
       "      <td>analyte</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ce309009-9257-4c13-9241-83fc629c35a0</td>\n",
       "      <td>f0003f0a-07ea-548e-b1f7-7e6d1b27d47a</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              identifier  \\\n",
       "0   4591a53d-5668-4a70-b44b-e08a3d59267e   \n",
       "1   c53c4d60-2ddb-5da8-932e-00a86fa2347f   \n",
       "2   31075cfa-7aef-59f1-bf54-a1cddb5ee0fd   \n",
       "3   0d8adcbf-13f0-48c3-83df-3fa205b79ae8   \n",
       "4   d085ebd9-7605-54a0-abb9-10867f5fa1b1   \n",
       "5   a31724b6-e550-552b-bd61-41341c534e28   \n",
       "6   9250d96e-1cdc-4d68-8a56-f7b186a6fab5   \n",
       "7   b12c257d-7409-4858-9384-c430929a075a   \n",
       "8   702d7ba0-9558-5b2d-af4d-cd797485b8c1   \n",
       "9   f0003f0a-07ea-548e-b1f7-7e6d1b27d47a   \n",
       "10  ce309009-9257-4c13-9241-83fc629c35a0   \n",
       "\n",
       "                   derived_from_specimen specimen_type       source_material  \n",
       "0                       initial specimen        sample         Primary Tumor  \n",
       "1   4591a53d-5668-4a70-b44b-e08a3d59267e       portion         Primary Tumor  \n",
       "2   c53c4d60-2ddb-5da8-932e-00a86fa2347f       analyte         Primary Tumor  \n",
       "3   31075cfa-7aef-59f1-bf54-a1cddb5ee0fd       aliquot         Primary Tumor  \n",
       "4   4591a53d-5668-4a70-b44b-e08a3d59267e       portion         Primary Tumor  \n",
       "5   d085ebd9-7605-54a0-abb9-10867f5fa1b1       analyte         Primary Tumor  \n",
       "6   a31724b6-e550-552b-bd61-41341c534e28       aliquot         Primary Tumor  \n",
       "7                       initial specimen        sample  Blood Derived Normal  \n",
       "8   b12c257d-7409-4858-9384-c430929a075a       portion  Blood Derived Normal  \n",
       "9   702d7ba0-9558-5b2d-af4d-cd797485b8c1       analyte  Blood Derived Normal  \n",
       "10  f0003f0a-07ea-548e-b1f7-7e6d1b27d47a       aliquot  Blood Derived Normal  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "rowsFromRoot = []\n",
    "for s in r[0]['ResearchSubject'][0]['Specimen']:\n",
    "    rowsFromRoot.append([s['identifier'][0]['value'], s['derived_from_specimen'], s['specimen_type'], s['source_material_type']])\n",
    "df = pd.DataFrame(rowsFromRoot)\n",
    "df.columns = ['identifier','derived_from_specimen','specimen_type','source_material']\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Then the PDC \"Research Subject\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>derived_from_specimen</th>\n",
       "      <th>specimen_type</th>\n",
       "      <th>source_material</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>f4af3e4d-641b-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>initial specimen</td>\n",
       "      <td>sample</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>208ebc64-6425-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>f4af3e4d-641b-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>f6cce507-641b-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>initial specimen</td>\n",
       "      <td>sample</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>08b2a2bf-6427-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>f6cce507-641b-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>44f5b956-642a-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>f6cce507-641b-11e8-bcf1-0a2705229b82</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>Primary Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             identifier                 derived_from_specimen  \\\n",
       "0  f4af3e4d-641b-11e8-bcf1-0a2705229b82                      initial specimen   \n",
       "1  208ebc64-6425-11e8-bcf1-0a2705229b82  f4af3e4d-641b-11e8-bcf1-0a2705229b82   \n",
       "2  f6cce507-641b-11e8-bcf1-0a2705229b82                      initial specimen   \n",
       "3  08b2a2bf-6427-11e8-bcf1-0a2705229b82  f6cce507-641b-11e8-bcf1-0a2705229b82   \n",
       "4  44f5b956-642a-11e8-bcf1-0a2705229b82  f6cce507-641b-11e8-bcf1-0a2705229b82   \n",
       "\n",
       "  specimen_type      source_material  \n",
       "0        sample  Solid Tissue Normal  \n",
       "1       aliquot  Solid Tissue Normal  \n",
       "2        sample        Primary Tumor  \n",
       "3       aliquot        Primary Tumor  \n",
       "4       aliquot        Primary Tumor  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rowsFromRoot = []\n",
    "for s in r[0]['ResearchSubject'][1]['Specimen']:\n",
    "    rowsFromRoot.append([s['identifier'][0]['value'], s['derived_from_specimen'], s['specimen_type'], s['source_material_type']])\n",
    "df = pd.DataFrame(rowsFromRoot)\n",
    "df.columns = ['identifier','derived_from_specimen','specimen_type','source_material']\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In respect of the PDC in v1 we got just two specimens, now we have five. This is good. Those additional specimens for the PDC ResearchSubject now have the distinct aliquots that went to PNNL and VU. That fixes an issue from v1. It was described as issue 4 in the original notebook and reported recently as reported as [cda-python/issues/97](https://github.com/CancerDataAggregator/cda-python/issues/97)\n",
    "\n",
    "If things are going well we would expect the VU and PNNL files (see the original notebook) only to be associated with the specific aliquot. From what I've seen elsewhere though I'm not confident that will be the case. We have seen that files have been included not only within specimens at higher levels in the tree, but at lower or peer level specimens. That is documented in [cda-service/issues/79](https://github.com/CancerDataAggregator/cda-service/issues/79).\n",
    "\n",
    "Pessimism aside let's try it! Given issue 79 let's restrict ourselves to the aliquots. Given the ground truth diagram we expect: \n",
    "one aliquot from the normal to have PNNL files only\n",
    "one aliquot from the tumor to have PNNL files only\n",
    "one aliquot from the tumor to have VU files only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id : 208ebc64-6425-11e8-bcf1-0a2705229b82\n",
      "identifier [{'system': 'PDC', 'value': '208ebc64-6425-11e8-bcf1-0a2705229b82'}]\n",
      "associated_project : CPTAC-2\n",
      "age_at_collection : None\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "anatomical_site : Not Reported\n",
      "source_material_type : Solid Tissue Normal\n",
      "specimen_type : aliquot\n",
      "derived_from_specimen : f4af3e4d-641b-11e8-bcf1-0a2705229b82\n",
      "derived_from_subject : 09CO022\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzid.gz\n",
      "________________________________________________________________________________\n",
      "id : 08b2a2bf-6427-11e8-bcf1-0a2705229b82\n",
      "identifier [{'system': 'PDC', 'value': '08b2a2bf-6427-11e8-bcf1-0a2705229b82'}]\n",
      "associated_project : CPTAC-2\n",
      "age_at_collection : None\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "anatomical_site : Not Reported\n",
      "source_material_type : Primary Tumor\n",
      "specimen_type : aliquot\n",
      "derived_from_specimen : f6cce507-641b-11e8-bcf1-0a2705229b82\n",
      "derived_from_subject : 09CO022\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw.cap.psm\n",
      "________________________________________________________________________________\n",
      "id : 44f5b956-642a-11e8-bcf1-0a2705229b82\n",
      "identifier [{'system': 'PDC', 'value': '44f5b956-642a-11e8-bcf1-0a2705229b82'}]\n",
      "associated_project : CPTAC-2\n",
      "age_at_collection : None\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "anatomical_site : Not Reported\n",
      "source_material_type : Primary Tumor\n",
      "specimen_type : aliquot\n",
      "derived_from_specimen : f6cce507-641b-11e8-bcf1-0a2705229b82\n",
      "derived_from_subject : 09CO022\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw.cap.psm\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "subject2 = r[0]['ResearchSubject'][1]\n",
    "for s2 in subject2['Specimen']:        \n",
    "    if s2['specimen_type'] == 'aliquot':\n",
    "        for k, v in s2.items():\n",
    "            if v.__class__.__name__ == 'list':\n",
    "                if len(v) < 2:\n",
    "                    print(k, v)\n",
    "                else:\n",
    "                    if k == 'File':\n",
    "                        for f in v:\n",
    "                            print(f['label'])\n",
    "                    else:    \n",
    "                        print ('{} has {} items'.format(k, len(v)))\n",
    "            else:\n",
    "                print ('{} : {}'.format(k, v))\n",
    "\n",
    "        print('_'*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bingo! That appears to be the answer expected! It's certainly a distinct set of files for each of the three specimens.\n",
    "\n",
    "What happens with files at the higher level?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "id : f4af3e4d-641b-11e8-bcf1-0a2705229b82\n",
      "identifier [{'system': 'PDC', 'value': 'f4af3e4d-641b-11e8-bcf1-0a2705229b82'}]\n",
      "associated_project : CPTAC-2\n",
      "age_at_collection : None\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "anatomical_site : Not Reported\n",
      "source_material_type : Solid Tissue Normal\n",
      "specimen_type : sample\n",
      "derived_from_specimen : initial specimen\n",
      "derived_from_subject : 09CO022\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw.cap.psm\n",
      "________________________________________________________________________________\n",
      "id : f6cce507-641b-11e8-bcf1-0a2705229b82\n",
      "identifier [{'system': 'PDC', 'value': 'f6cce507-641b-11e8-bcf1-0a2705229b82'}]\n",
      "associated_project : CPTAC-2\n",
      "age_at_collection : None\n",
      "primary_disease_type : Colon Adenocarcinoma\n",
      "anatomical_site : Not Reported\n",
      "source_material_type : Primary Tumor\n",
      "specimen_type : sample\n",
      "derived_from_specimen : initial specimen\n",
      "derived_from_subject : 09CO022\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzid.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f03.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f09.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f06.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f03.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f10.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.raw\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f01.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw.cap.psm\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f02.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f08.raw.cap.psm\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f04.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f07.raw\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f12.mzML.gz\n",
      "13CPTAC_COprospective_P_PNNL_20170215_B4S1_f01.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f05.mzML.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f11.mzid.gz\n",
      "13CPTAC_COprospective_W_PNNL_20170123_B4S1_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f09.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f03.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.mzid.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f02.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.mzid.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f01.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f02.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f05.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f11.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f04.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f10.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f06.mzML.gz\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f05.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw.cap.psm\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f04.raw.cap.psm\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f01.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f07.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f06.mzid.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f08.mzML.gz\n",
      "16CPTAC_COprospective_W_PNNL_20170123_B4S4_f12.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.mzML.gz\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f05.raw\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f03.raw\n",
      "16CPTAC_COprospective_P_PNNL_20170215_B4S4_f03.raw.cap.psm\n",
      "87CPTAC_COprospective_W_VU_20160701_09CO022_f02.raw.cap.psm\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "for s2 in subject2['Specimen']:        \n",
    "    if s2['specimen_type'] == 'sample':\n",
    "        for k, v in s2.items():\n",
    "            if v.__class__.__name__ == 'list':\n",
    "                if len(v) < 2:\n",
    "                    print(k, v)\n",
    "                else:\n",
    "                    if k == 'File':\n",
    "                        for f in v:\n",
    "                            print(f['label'])\n",
    "                    else:    \n",
    "                        print ('{} has {} items'.format(k, len(v)))\n",
    "            else:\n",
    "                print ('{} : {}'.format(k, v))\n",
    "\n",
    "        print('_'*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That there are two specimens at this level is as expected. The tumor and the normal.\n",
    "\n",
    "However the files lists for each specimen are dangerously inaccurate. Under the normal specimen it lists files that came from the tumor. The reverse is also true - under the tumor specimen are listed files which are derived from the normal.\n",
    "\n",
    "A further independent check of this can be made by searches in the PDC. That indicates no work was done on at VU on the normal sample. \n",
    "\n",
    "This has been submitted as [cda-python/issues/99](https://github.com/CancerDataAggregator/cda-python/issues/99)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
