{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "331faaed",
   "metadata": {},
   "source": [
    "### Retrieving files at a specific level within a Subject\n",
    "\n",
    "Issue [cda-python/issues/93](https://github.com/CancerDataAggregator/cda-python/issues/93) raised the need to retrieve files at a specific level within a Subject's hierarchy of data.\n",
    "\n",
    "This is an illustration of how Q.sql() can be used to do so.\n",
    "\n",
    "An unintended consequence of this exploration was that it turned up another consequence of the repetition (> than duplication) of files redundantly within the specimen hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a599c4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cdapython import Q\n",
    "import json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b36b3468",
   "metadata": {},
   "source": [
    "### A query to get file details from the Research Subject level\n",
    "To keep the output compact we're limiting our query to a specific Subject and to BAM files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e70a7fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "QueryID: c8e8c40c-e6bd-46d1-aa74-23a32c2c182e\n",
       "Query: SELECT su.id sid , rs.identifier rsid, fi.drs_uri\n",
       "from gdc-bq-sample.integration.all_v2 AS su,\n",
       "unnest(ResearchSubject) AS rs,\n",
       "unnest(rs.File) as fi\n",
       "where (su.id = 'TCGA-13-1504')\n",
       "and file_format = 'BAM' \n",
       "Offset: 0\n",
       "Count: 3\n",
       "Total Row Count: 3\n",
       "More pages: False"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query1 = \"\"\"SELECT su.id sid , rs.identifier rsid, fi.drs_uri\n",
    "from gdc-bq-sample.integration.all_v2 AS su,\n",
    "unnest(ResearchSubject) AS rs,\n",
    "unnest(rs.File) as fi\n",
    "where (su.id = 'TCGA-13-1504')\n",
    "and file_format = 'BAM' \"\"\"\n",
    "\n",
    "r1 = Q.sql(query1)\n",
    "r1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a73e8eb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "for row in r1:\n",
    "    print(json.dumps(row, indent=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d882fb19",
   "metadata": {},
   "source": [
    "### Or to provide more complete details of the file\n",
    "We specify all fields for the File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fff6702f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"8b7859ed-395a-449d-8e4b-7d385d0ffa43\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"TCGA-13-1504-01A-01R-1565-13_mirna_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\",\n",
      "   \"byte_size\": \"617903851\",\n",
      "   \"checksum\": \"42b51bf61122b716cd02a628da9b7d89\"\n",
      "}\n",
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"698994fe-b22a-4723-b9f0-23551fd3bb0e\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"C239.TCGA-13-1504-01A-01W.5_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\",\n",
      "   \"byte_size\": \"32100267850\",\n",
      "   \"checksum\": \"81412db2d755379b4d0d24c860013ec2\"\n",
      "}\n",
      "{\n",
      "   \"sid\": \"TCGA-13-1504\",\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"07a9e595-023d-47d5-97fa-3fcc62f5d0c6\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"C239.TCGA-13-1504-10A-01W.6_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\",\n",
      "   \"byte_size\": \"35466456040\",\n",
      "   \"checksum\": \"74451615312647e65cc6e6e69c0a9e0b\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "query2 = \"\"\"SELECT su.id sid , rs.identifier rsid, fi.*\n",
    "from gdc-bq-sample.integration.all_v2 AS su,\n",
    "unnest(ResearchSubject) AS rs,\n",
    "unnest(rs.File) as fi\n",
    "where (su.id = 'TCGA-13-1504')\n",
    "and file_format = 'BAM' \"\"\"\n",
    "\n",
    "r2 = Q.sql(query2)\n",
    "for row in r2:\n",
    "    print(json.dumps(row, indent=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7635c43",
   "metadata": {},
   "source": [
    "Note that some care is needed in reading the results. The identifier attribute belongs to the File, as do all the subsequent attributes within each result row. Maybe there is a way of writing the select clause to retain this."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0ddc43",
   "metadata": {},
   "source": [
    "### Digging deeper - File details from the specimen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "62d69b42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "QueryID: 777537e6-861b-446a-a920-a254913ee66c\n",
       "Query: SELECT rs.identifier rsid, fi.drs_uri\n",
       "from gdc-bq-sample.integration.all_v2 AS su,\n",
       "unnest(ResearchSubject) AS rs,\n",
       "unnest(rs.Specimen) AS sp,\n",
       "unnest(sp.File) as fi\n",
       "where (su.id = 'TCGA-13-1504')\n",
       "and fi.file_format = 'BAM' \n",
       "Offset: 0\n",
       "Count: 16\n",
       "Total Row Count: 16\n",
       "More pages: False"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specimenFileQuery = \"\"\"SELECT rs.identifier rsid, fi.drs_uri\n",
    "from gdc-bq-sample.integration.all_v2 AS su,\n",
    "unnest(ResearchSubject) AS rs,\n",
    "unnest(rs.Specimen) AS sp,\n",
    "unnest(sp.File) as fi\n",
    "where (su.id = 'TCGA-13-1504')\n",
    "and fi.file_format = 'BAM' \"\"\"\n",
    "specimenFiles = Q.sql(specimenFileQuery)\n",
    "specimenFiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b5fdf70",
   "metadata": {},
   "source": [
    "### That's odd\n",
    "We had three BAM files for this subject before. Now we have sixteen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "626fdecf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "for row in specimenFiles:\n",
    "    print(json.dumps(row, indent=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b15bc549",
   "metadata": {},
   "source": [
    "Close inspection of the above shows that there are really only three files present.\n",
    "\n",
    "### What's going on?\n",
    "Some inspection showed where the problem was. The following query illustrates by adding the specimen type to the select statement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c1776e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"specimen_type\": \"sample\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"portion\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"analyte\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"aliquot\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"sample\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"sample\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"portion\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"portion\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"slide\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"slide\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"slide\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"slide\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"analyte\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"aliquot\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"analyte\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n",
      "{\n",
      "   \"specimen_type\": \"aliquot\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "specimenFileQuery2 = \"\"\"SELECT specimen_type, fi.drs_uri\n",
    "from gdc-bq-sample.integration.all_v2 AS su,\n",
    "unnest(ResearchSubject) AS rs,\n",
    "unnest(rs.Specimen) AS sp,\n",
    "unnest(sp.File) as fi\n",
    "where (su.id = 'TCGA-13-1504')\n",
    "and fi.file_format = 'BAM' \"\"\"\n",
    "specimenFiles2 = Q.sql(specimenFileQuery2)\n",
    "for row in specimenFiles2:\n",
    "    print(json.dumps(row, indent=3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f2eeda",
   "metadata": {},
   "source": [
    "#### Diagnosis\n",
    "The problem derives from the fact that the File has been repeated as part of the content of multiple specimens. This was the duplication referred to in the v2 updates recently added to [cda-service/issues/79](https://github.com/CancerDataAggregator/cda-service/issues/79).\n",
    "\n",
    "### Fixing a hole\n",
    "We can work around the problem by specifying the specific specimen type (aliquot) in whose files we are interested in. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8d8d6a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "QueryID: 0330f456-55cc-495f-bad4-e5eafa0898bb\n",
       "Query: SELECT rs.identifier rsid, fi.*\n",
       "from gdc-bq-sample.integration.all_v2 AS su,\n",
       "unnest(ResearchSubject) AS rs,\n",
       "unnest(rs.Specimen) AS sp,\n",
       "unnest(sp.File) as fi\n",
       "where (su.id = 'TCGA-13-1504')\n",
       "and specimen_type = 'aliquot'\n",
       "and fi.file_format = 'BAM' \n",
       "Offset: 0\n",
       "Count: 3\n",
       "Total Row Count: 3\n",
       "More pages: False"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specimenFileQuery3 = \"\"\"SELECT rs.identifier rsid, fi.*\n",
    "from gdc-bq-sample.integration.all_v2 AS su,\n",
    "unnest(ResearchSubject) AS rs,\n",
    "unnest(rs.Specimen) AS sp,\n",
    "unnest(sp.File) as fi\n",
    "where (su.id = 'TCGA-13-1504')\n",
    "and specimen_type = 'aliquot'\n",
    "and fi.file_format = 'BAM' \"\"\"\n",
    "specimenFiles3 = Q.sql(specimenFileQuery3)\n",
    "specimenFiles3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902bf693",
   "metadata": {},
   "source": [
    "That gets us back to the expected number of files i.e. with no repetition.\n",
    "\n",
    "And we can list the results as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0bf4f500",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"07a9e595-023d-47d5-97fa-3fcc62f5d0c6\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"07a9e595-023d-47d5-97fa-3fcc62f5d0c6\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"C239.TCGA-13-1504-10A-01W.6_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:07a9e595-023d-47d5-97fa-3fcc62f5d0c6\",\n",
      "   \"byte_size\": \"35466456040\",\n",
      "   \"checksum\": \"74451615312647e65cc6e6e69c0a9e0b\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"698994fe-b22a-4723-b9f0-23551fd3bb0e\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"698994fe-b22a-4723-b9f0-23551fd3bb0e\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"C239.TCGA-13-1504-01A-01W.5_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:698994fe-b22a-4723-b9f0-23551fd3bb0e\",\n",
      "   \"byte_size\": \"32100267850\",\n",
      "   \"checksum\": \"81412db2d755379b4d0d24c860013ec2\"\n",
      "}\n",
      "{\n",
      "   \"rsid\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"cd49126a-ec15-43fa-9e43-3f7460d43f2b\"\n",
      "      }\n",
      "   ],\n",
      "   \"id\": \"8b7859ed-395a-449d-8e4b-7d385d0ffa43\",\n",
      "   \"identifier\": [\n",
      "      {\n",
      "         \"system\": \"GDC\",\n",
      "         \"value\": \"8b7859ed-395a-449d-8e4b-7d385d0ffa43\"\n",
      "      }\n",
      "   ],\n",
      "   \"label\": \"TCGA-13-1504-01A-01R-1565-13_mirna_gdc_realn.bam\",\n",
      "   \"data_category\": \"Sequencing Reads\",\n",
      "   \"data_type\": \"Aligned Reads\",\n",
      "   \"file_format\": \"BAM\",\n",
      "   \"associated_project\": \"TCGA-OV\",\n",
      "   \"drs_uri\": \"drs://dg.4DFC:8b7859ed-395a-449d-8e4b-7d385d0ffa43\",\n",
      "   \"byte_size\": \"617903851\",\n",
      "   \"checksum\": \"42b51bf61122b716cd02a628da9b7d89\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "for row in specimenFiles3:\n",
    "    print(json.dumps(row, indent=3))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
